libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/WaymarkOverlay.kt

```diff
@@ -0,0 +1,159 @@
+package net.cyclestreets.views.overlay
+
+import android.content.SharedPreferences
+import android.content.SharedPreferences.Editor
+import android.graphics.Canvas
+import android.graphics.Matrix
+import android.graphics.Paint
+import android.graphics.Point
+import android.graphics.drawable.BitmapDrawable
+import android.graphics.drawable.Drawable
+import android.support.v4.content.res.ResourcesCompat
+
+import net.cyclestreets.routing.Journey
+import net.cyclestreets.routing.Route
+import net.cyclestreets.routing.Waypoints
+import net.cyclestreets.view.R
+import net.cyclestreets.views.CycleMapView
+
+import org.osmdroid.api.IGeoPoint
+import org.osmdroid.api.IProjection
+import org.osmdroid.util.GeoPoint
+import org.osmdroid.views.MapView
+import org.osmdroid.views.overlay.Overlay
+import org.osmdroid.views.overlay.OverlayItem
+
+import java.util.ArrayList
+
+class WaymarkOverlay(private val mapView: CycleMapView) : Overlay(), PauseResumeListener, Route.Listener {
+
+    private val greenWisp: Drawable?
+    private val orangeWisp: Drawable?
+    private val redWisp: Drawable?
+    private val screenPos = Point()
+    private val bitmapTransform = Matrix()
+    private val bitmapPaint = Paint()
+
+    private val waymarkers: MutableList<OverlayItem>
+
+    init {
+        val res = mapView.context.resources
+        val makeWisp = { drawable: Int -> ResourcesCompat.getDrawable(res, drawable, null) }
+
+        greenWisp = makeWisp(R.drawable.green_wisp)
+        orangeWisp = makeWisp(R.drawable.orange_wisp)
+        redWisp = makeWisp(R.drawable.red_wisp)
+
+        waymarkers = ArrayList()
+    }
+
+    fun waymarkersCount(): Int {
+        return waymarkers.size
+    }
+
+    fun waypoints(): Waypoints {
+        return Waypoints(waymarkers.map { wp -> wp.point })
+    }
+
+    fun finish(): IGeoPoint {
+        return waymarkers.last().point
+    }
+
+    fun addWaypoint(point: IGeoPoint?) {
+        if (point == null)
+            return
```

Comment:
I'm trying to figure out under what circumstances `point` would be `null` and I can't think of any.

Certainly, where it's now called from `tapAction()` in the other `Overlay`, the parameter passed in is non-nullable (so if it is null we'll fail at a higher level than this.)

Maybe we could just make this non-nullable and then see what happens when we play with the app...?

Reply:
I wondered too.  I tried to remembered if that's there because it did happen, or due to an abundance of over caution.  I'll give it a go.